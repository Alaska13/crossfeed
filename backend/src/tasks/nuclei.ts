import { Domain, Vulnerability } from '../models';
import { CommandOptions } from './ecs-client';
import getLiveWebsites from './helpers/getLiveWebsites';
import PQueue from 'p-queue';
import { spawnSync } from 'child_process';
import saveVulnerabilitiesToDb from './helpers/saveVulnerabilitiesToDb';
import { plainToClass } from 'class-transformer';

interface NucleiResult {
  template: string;
  type: string;
  matched: string;
  severity: string;
  author: string;
  description: string;
}

/** Maps Nuclei severity to severity stored in the Vulnerabilities database.
 * See all severities at https://github.com/projectdiscovery/nuclei/blob/bc320d76fb5c19e9c958204cc0d72365b87106a4/internal/runner/templates.go#L19-L25.
 * TODO: it might be better to instead get this information from the NVD database.
 */
const SEVERITY_MAP = {
  info: 'None',
  low: 'Low',
  medium: 'Medium',
  high: 'High',
  critical: 'Critical'
};

const nuclei = async (domain: Domain): Promise<void> => {
  console.log('Domain', domain.name);
  for (const service of domain.services) {
    if (service.port !== 80 && service.port !== 443) {
      continue;
    }
    const url =
      service.port === 443 ? `https://${domain.name}` : `http://${domain.name}`;
    // TODO: consider running templates under nuclei-templates/vulnerabilities as well (those templates aren't
    // directly tied to CVEs, though).
    const { stdout, stderr, status } = spawnSync('nuclei', [
      '-target',
      url,
      '-t',
      'nuclei-templates/cves',
      '-json',
      '-silent',
      '-proxy-url',
      process.env.GLOBAL_AGENT_HTTP_PROXY!
    ]);
    if (stderr.toString()) {
      console.error('stderr', stderr.toString());
    }
    if (status !== 0) {
      console.error('Nuclei failed');
      continue;
    }
    const output = stdout.toString();
    const results: NucleiResult[] = output
      .split('\n')
      .filter((e) => e)
      .map((e) => JSON.parse(e));
    service.nucleiResults = results;
    await service.save();
    const vulnerabilities = results
      .filter((e) => e.template.startsWith('CVE-'))
      .map((e) =>
        plainToClass(Vulnerability, {
          domain: domain,
          lastSeen: new Date(Date.now()),
          title: e.template,
          cve: e.template,
          severity: SEVERITY_MAP[e.severity],
          state: 'open',
          substate: 'exploitable'
        })
      );
    await saveVulnerabilitiesToDb(vulnerabilities, false);
  }
};

export const handler = async (commandOptions: CommandOptions) => {
  const { organizationId, organizationName } = commandOptions;

  console.log('Running nuclei on organization', organizationName);

  const liveWebsites = await getLiveWebsites(organizationId!);
  const queue = new PQueue({ concurrency: 2 });
  await Promise.all(liveWebsites.map((site) => queue.add(() => nuclei(site))));

  console.log(`Nuclei finished for ${liveWebsites.length} domains`);
};
