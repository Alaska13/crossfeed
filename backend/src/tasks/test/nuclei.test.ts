import { handler as nuclei } from '../nuclei';
import {
  connectToDatabase,
  Organization,
  Service,
  Domain,
  Vulnerability
} from '../../models';
import { spawnSync } from 'child_process';

jest.mock('child_process', () => ({
  spawnSync: jest.fn()
}));

const mockNucleiResponse = (response) => {
  (spawnSync as jest.Mock).mockImplementation(() => ({
    status: 0,
    stderr: '',
    stdout: response
  }));
};

describe('nuclei', () => {
  beforeAll(async () => {
    await connectToDatabase();
  });
  let organization;
  beforeEach(async () => {
    organization = await Organization.create({
      name: 'test-' + Math.random(),
      rootDomains: ['test-' + Math.random()],
      ipBlocks: [],
      isPassive: false
    }).save();
  });
  test('basic test', async () => {
    mockNucleiResponse(`{"template":"CVE-2018-5230","type":"http","matched":"https://google-gruyere.appspot.com/123456//pages/includes/status-list-mo%3CIFRAME%20SRC%3D%22javascript%3Aalert%281337%29%22%3E.vm","severity":"medium","author":"madrobot","description":""}
{"template":"CVE-2018-1000129","type":"http","matched":"https://google-gruyere.appspot.com/123456//api/jolokia/read%3Csvg%20onload=alert%28document.domain%29%3E?mimeType=text/html","severity":"high","author":"mavericknerd @0h1in9e","description":"An XSS vulnerability exists in the Jolokia agent version 1.3.7 in the HTTP servlet that allows an attacker to execute malicious javascript in the victim's browser."}
{"template":"CVE-2018-1000129","type":"http","matched":"https://google-gruyere.appspot.com/123456/:8080/jolokia/read%3Csvg%20onload=alert%28document.domain%29%3E?mimeType=text/html","severity":"high","author":"mavericknerd @0h1in9e","description":"An XSS vulnerability exists in the Jolokia agent version 1.3.7 in the HTTP servlet that allows an attacker to execute malicious javascript in the victim's browser."}`);
    const domain = await Domain.create({
      organization,
      name: 'google-gruyere.appspot.com',
      ip: '0.0.0.0'
    }).save();
    let service = await Service.create({
      service: 'https',
      port: 443,
      domain
    }).save();
    await nuclei({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: 'scanId',
      scanName: 'scanName',
      scanTaskId: 'scanTaskId',
      chunkNumber: 0,
      numChunks: 1
    });
    service = await Service.findOneOrFail(service.id);
    expect(service.nucleiResults).toMatchSnapshot();
    const vulnerabilities = await Vulnerability.find({ domain });
    expect(vulnerabilities.length).toEqual(2);
    expect(vulnerabilities[0]).toMatchSnapshot({
      createdAt: expect.any(Date),
      lastSeen: expect.any(Date),
      id: expect.any(String)
    });
  });
});
