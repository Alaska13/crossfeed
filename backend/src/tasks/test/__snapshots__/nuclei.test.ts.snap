// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nuclei basic test 1`] = `
Array [
  Object {
    "author": "madrobot",
    "description": "",
    "matched": "https://google-gruyere.appspot.com/123456//pages/includes/status-list-mo%3CIFRAME%20SRC%3D%22javascript%3Aalert%281337%29%22%3E.vm",
    "severity": "medium",
    "template": "CVE-2018-5230",
    "type": "http",
  },
  Object {
    "author": "mavericknerd @0h1in9e",
    "description": "An XSS vulnerability exists in the Jolokia agent version 1.3.7 in the HTTP servlet that allows an attacker to execute malicious javascript in the victim's browser.",
    "matched": "https://google-gruyere.appspot.com/123456//api/jolokia/read%3Csvg%20onload=alert%28document.domain%29%3E?mimeType=text/html",
    "severity": "high",
    "template": "CVE-2018-1000129",
    "type": "http",
  },
  Object {
    "author": "mavericknerd @0h1in9e",
    "description": "An XSS vulnerability exists in the Jolokia agent version 1.3.7 in the HTTP servlet that allows an attacker to execute malicious javascript in the victim's browser.",
    "matched": "https://google-gruyere.appspot.com/123456/:8080/jolokia/read%3Csvg%20onload=alert%28document.domain%29%3E?mimeType=text/html",
    "severity": "high",
    "template": "CVE-2018-1000129",
    "type": "http",
  },
]
`;

exports[`nuclei basic test 2`] = `
Object {
  "cpe": null,
  "createdAt": Any<Date>,
  "cve": "CVE-2018-1000129",
  "cvss": null,
  "cwe": null,
  "description": "",
  "id": Any<String>,
  "lastSeen": Any<Date>,
  "notes": "",
  "severity": "High",
  "state": "open",
  "substate": "exploitable",
  "title": "CVE-2018-1000129",
  "updatedAt": 2020-09-24T21:41:05.747Z,
}
`;
